libqtile/widget/wallpaper.py

```diff
@@ -53,22 +55,26 @@ def __init__(self, **config):
 
         self.add_callbacks({'Button1': self.set_wallpaper})
 
-    def _configure(self, qtile, bar):
-        base._TextBox._configure(self, qtile, bar)
-        if not self.bar.screen.wallpaper:
-            self.set_wallpaper()
+    def poll(self):
+        """
+        Return the current wallpaper
+        """
+        self.set_wallpaper()
+        return self.text
 
     def get_path(self, file):
         return os.path.join(os.path.expanduser(self.directory), file)
 
     def get_wallpapers(self):
         try:
             # get path of all files in the directory
-            self.images = list(
-                filter(os.path.isfile,
-                       map(self.get_path,
-                           os.listdir(
-                               os.path.expanduser(self.directory)))))
+            for root, dirnames, filenames in os.walk(os.path.expanduser(self.directory)):
```

Comment:
What about svg, xcf, otf, etc? Doesn't this break people who were using those formats? I don't see any reason to change this behavior in this patch.

Reply:
Similar scanning-folder-for-files logic has been written in `libqtile.images` (function `get_matching_files`). Perhaps we should move that more general logic into `libqtile.utils`, and then the widget can use that. Indeed, if whichever 'general' implementation returns something simple like a list of paths, then the widget only needs to accept a list of paths. No need to hard-code any extensions, as `libqtile.images.get_matching_files` already uses a regex, and the user can then decide what to use, and we can just log errors.

Reply:
For the extension filter, normally when we pull some wallpaper from website, there will be files like README, and it'll not be rendered, thus we need to filter them. svg, xcf are just vector files and of course you can use it as wallpaper, but 99% of people will just stick with jpg and png afaik.
For the common functions, I didn't know it exists...will check later.

Reply:
The libqtile.images module is not on documentation...
http://docs.qtile.org/en/latest/_modules/index.html
That's why I didn't know it.

Reply:
For get_matching_files,
after reading the code, currently it's used only by getting the icons. So maybe stick to my implementation for the widget now?

Reply:
> The libqtile.images module is not in the documentation

I think this is because it is used only internally, so isn't a 'user-facing' module. No probs though. I implemented the more general file-scanning utility function so you can hook into that, which might keep things simple. If you check out the `scan_files` function in `libqtile.utils` from PR #2022, you can see how to use it. What this would allow for is receiving a list of glob strings as arguments to the widget that would be used to find the appropriate image files. This could then replace the hard-coded `fnmatch.filter` lines so users can use whatever image type they use. Any thoughts? :)

Reply:
That scan_files function has been merged so you could use it for your widget. It is [here](https://github.com/qtile/qtile/blob/master/libqtile/utils.py#L301)